### PR TITLE
Unknown Version Handling

### PR DESCRIPTION
**Pull request by vkrizan**
Adds templating support (with `-o` argument) in case when a version of toplevel or vendored package is not matched. Unknown version is by default marked with `?`.

Example (with unknown toplevel version):
```
github.com/rogpeppe/godef:?
github.com/rogpeppe/godef:?/9fans.net/go:v0.0.0-0.20181112111441-237454027057
github.com/rogpeppe/godef:?/golang.org/x/tools:v0.0.0-0.20181130195746-895048a75ecf
```

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -101,9 +101,6 @@ func showTopLevel(tmpl *template.Template, src *retrodep.GoSource) *retrodep.Ref
 		display(tmpl, topLevelMarker, project)
 	default:
 		log.Fatalf("%s: %s", src.Path, err)
-	}
-
-	if err != nil {
 		return nil
```
Comment (twaugh) [inline main.go]: `log.Fatalf` [will call](https://godoc.org/log#Fatalf) `os.Exit(1)` so we'll never get to this `return nil` line. We should remove that too.

Other than that -- nice catch, I think this was a bug before.

File: main.go [hunk 2]
```diff
@@ -47,8 +47,12 @@ var exitFirst = flag.Bool("x", false, "exit on the first failure")
 var errorShown = false
 var usage func(string)
 
-func displayUnknown(name string) {
-	fmt.Printf("%s ?\n", name)
+func displayUnknown(tmpl *template.Template, name string, ref *retrodep.Reference, root string) {
+	if ref == nil || *templateArg != "" {
+		fmt.Printf("%s%s ?v\n", name, root)
```
Comment (twaugh) [inline main.go]: Looks like there is a stray 'v' here.

File: retrodep/vendored.go [hunk 3]
```diff
@@ -343,6 +343,13 @@ func (src GoSource) DescribeProject(
 		topver = top.Ver
 	}
 
+	ref = &Reference{
+		TopPkg: toppkg,
+		TopVer: topver,
+		Pkg:    project.Root,
+		Repo:   project.Repo,
+	}
+
```
Comment (twaugh) [inline retrodep/vendored.go:352]: Maybe now we can just update this `ref` to set the other fields, in the other places we currently create a new one?

File: main.go [hunk 4]
```diff
@@ -30,7 +30,8 @@ import (
 	"github.com/release-engineering/retrodep/retrodep"
 )
 
-const defaultTemplate string = "{{if .TopPkg}}{{.TopPkg}}:{{.TopVer}}/{{end}}{{.Pkg}}:{{.Ver}}"
+const defaultTemplate string = "{{if .TopPkg}}{{.TopPkg}}:{{if .TopVer}}{{.TopVer}}{{else}}vX.X.X{{end}}" +
+	"/{{end}}{{.Pkg}}:{{if .Ver}}{{.Ver}}{{else}}vX.X.X{{end}}"
```
Comment (twaugh) [inline main.go]: I think I'd like something shorter than `vX.X.X`. Is `?` no good there?

**Comment by twaugh** [comment]
Thanks! Added some comments.

Also, in #99 I've added some tests for template handling to make sure it works the way we want it to.

One final thing: I've created a v0 branch. If we want to remove compatibility handling of `-template` from the master branch we're free to do that now. It will likely make the code simpler.

I quite like the way the version-match error is communicated in this PR -- just the absence of the version. Neat.

File: main.go (hunk 4 above)
Comment (alexandrevicenzi) [reply main.go]: Not sure, but could this be a multi-line string?

Comment (twaugh) [reply main.go]: > Not sure, but could this be a multi-line string?

You mean with backticks? Yes, I guess it could. We want to avoid newlines in the template output, so we'll need to use space trimming. I also just saw that we could use the "or" template function to make it simpler, e.g.
```
const defaultTemplate = `
  {{- if .TopPkg -}}
    {{.TopPkg}}:{{or .TopVer "?"}}/
  {{- end -}}
  {{.Pkg}}:{{or .Ver "?"}}`
```

Comment (vkrizan) [reply main.go]: Neat shortening with the `or` command. I guess we can remove the if for the toplevel package presence, as it is expected to be always present. Or we can use `{{or .TopPkg "UnidentifiedTopPackage"}}`. That would ensure to correctly identify a vendored package. But for that we would have to distinguish what is vendored and what is top-level only pkg (line).

How do you do the space trimming for such multiline string?

As for `vX.X.X` to `?`, I have no stronger preference. I just found `vX.X.X` more visually appealing.

Comment (twaugh) [reply main.go]: No, [Reference](https://godoc.org/github.com/release-engineering/retrodep/retrodep#Reference)'s TopPkg field will be empty if Pkg is the top-level package. That's how we arrange for packages reporting on only the top-level package to look different to those reporting on vendored dependencies.

The "{{- " and " -}}" bits do the [space-trimming](https://godoc.org/text/template#hdr-Text_and_spaces).

I quite like "?" as the 'unknown' value because it cannot be ether a valid semantic version or a valid tag name.

(While we're talking about the format, I would have preferred ";" as the separate between the top-level package detalis and the vendored package details, as "/" will appear in both names...)

File: main.go (hunk 1 above)
Comment (vkrizan) [reply main.go]: Should I keep the Fatal log and remove return, or, should I replace it with Error log instead?

File: retrodep/vendored.go [hunk 5]
```diff
@@ -397,7 +404,6 @@ func (src GoSource) DescribeProject(
 			Repo:   project.Repo,
 			Tag:    match,
 			Rev:    rev,
-			Ver:    match,
```
Comment (vkrizan) [inline retrodep/vendored.go:398]: Oh here it is. A bug. :smiling_imp:

File: main.go (hunk 2 above)
Comment (vkrizan) [reply main.go]: Fixed.

File: retrodep/vendored.go (hunk 3 above)
Comment (vkrizan) [reply retrodep/vendored.go:352]: Done. Thanks for the suggestion.

File: main.go (hunk 4 above)
Comment (vkrizan) [reply main.go]: I've use multiline with backticks and `?` for unknown version, as you suggested @twaugh.

As for the separator, I agree that we should change it. It would be easily parse-able. Lets do that in separate PR after this is merged.

File: retrodep/vendored.go (hunk 5 above)
Comment (vkrizan) [reply retrodep/vendored.go:398]: Fixed.

File: main.go [hunk 6]
```diff
@@ -47,8 +47,12 @@ var exitFirst = flag.Bool("x", false, "exit on the first failure")
 var errorShown = false
 var usage func(string)
 
-func displayUnknown(name string) {
-	fmt.Printf("%s ?\n", name)
+func displayUnknown(tmpl *template.Template, name string, ref *retrodep.Reference, root string) {
```
Comment (twaugh) [inline main.go]: Confused about what the parameters are here. What's name, and what's root?

I'd expect "topLevelMarker, name string", since that's what's getting passed in -- the optional "*" marker, and the name of the package. BUT: we never showed the topLevelMarker before, so there is no need to do that. Just `"%s ?\n", name` will keep compatibility, as long as name is the name of the package.

File: main.go [hunk 7]
```diff
@@ -47,8 +47,12 @@ var exitFirst = flag.Bool("x", false, "exit on the first failure")
...
+	if ref == nil || *templateArg != "" {
+		fmt.Printf("%s%s ?\n", name, root)
```
Comment (twaugh) [inline main.go]: I think this can just be `"%s ?\n", name` if name is the name of the package.

File: main.go (hunk 1 above)
Comment (twaugh) [reply main.go]: Keep `log.Fatalf` and remove `return nil` I think. We don't propagate errors from this function so we should handle them (by terminating).

File: main.go (hunk 6 above)
Comment (vkrizan) [reply main.go]: The `name` should have been `topLevelMarker`. The signature of `displayUnknown()` was reciprocated from `display()` (where the same contents to name are passed), only the `root` was added to make it compatible.

I'll see what can I do, to make it more clear.

Comment (vkrizan) [reply main.go]: I've changed the signature to:
```
displayUnknown(tmpl *template.Template, topLevelMarker string, ref *retrodep.Reference, projectRoot string) 
```
See 83baca8 and e4c3042.

File: main.go (hunk 7 above)
Comment (vkrizan) [reply main.go]: No, it cannot. `topLevelMarker` and `projectRoot` must be distinguished to be able to call `display()` correctly, and also for compatiblity. Please note, that the `projectRoot` is there for a case when the `ref` is nil.

File: main.go (hunk 1 above)
Comment (vkrizan) [reply main.go]: Return removed.

**Comment by twaugh** [comment]
Looks great! Would you mind re-basing it onto master (and maybe adding relevant tests)?

**Comment by vkrizan** [comment]
@twaugh rebased. I am not sure if I'd be able to write tests for these changes. Can you please help me with that?

**Comment by twaugh** [comment]
See what you think of this approach: https://github.com/release-engineering/retrodep/commit/2ea78a3e491513e7cecd83439279da8f8f1501bf

Alternative is to have those display functions write to an io.Writer. Then we can use os.Stdout in the real program, but pass in something we can inspect in the unit tests.

**Comment by vkrizan** [comment]
@twaugh Test at 2ea78a3 looks good to me. Should I incorporate it here, and add more cases, or, should we leave it for another PR?

I am not sure if I'd be able to rework the display functions to io.Writer at this point.

**Comment by twaugh** [comment]
Could you please incorporate it and add more cases (might just be one more, not sure).

**Comment by vkrizan** [comment]
@twaugh Added yours, as 263e2ff, and added one more test case (with non-zero templateArg) 80fbc49.

File: main_test.go [hunk 8]
```diff
@@ -15,7 +15,82 @@
...
+		},
+	}
+
+	for _, tc := range tcs {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			*templateArg = tc.templateArg
+			r, reset := captureStdout(t)
+			displayUnknown(nil, "*", nil, "example.com/foo")
```
Comment (twaugh) [inline main_test.go]: Oh, think this needs to use `tc.ref`.

Comment (vkrizan) [reply main_test.go]: Passed to it. 30683e8

**Review by twaugh — APPROVED** [review]
Nice!

**Review by athos-ribeiro — APPROVED** [review]
LGTM

**Comment by vkrizan** [comment]
Thank you Tim and Athos!
